pkg/dockerregistry/server/metrics/metrics.go

```diff
@@ -60,24 +94,61 @@ func (m *metricTimer) Stop() {
 	m.collector.WithLabelValues(m.labels...).Observe(float64(time.Since(m.startTime)) / float64(time.Second))
 }
 
-func newWrapper(reponame string) wrapped.Wrapper {
-	return func(ctx context.Context, funcname string, f func(ctx context.Context) error) error {
-		defer NewTimer(registryAPIRequests, []string{strings.ToLower(funcname), reponame}).Stop()
+// NewRepository wraps a distribution.Repository to collect statistics.
+func NewRepository(r distribution.Repository, reponame string) distribution.Repository {
+	return wrapped.NewRepository(r, func(ctx context.Context, funcname string, f func(ctx context.Context) error) error {
+		defer NewTimer(requestDurationSeconds, []string{strings.ToLower(funcname), reponame}).Stop()
 		return f(ctx)
-	}
+	})
+}
+
+// NewPullthroughRepository wraps distribution.Repository that is instantiated
+// for a remote registry to collect statistics.
+func NewPullthroughRepository(r distribution.Repository, registry string) distribution.Repository {
+	return wrapped.NewRepository(r, func(ctx context.Context, funcname string, f func(ctx context.Context) error) error {
+		labelValues := []string{strings.ToLower(registry), strings.ToLower(funcname)}
+		defer NewTimer(pullthroughRepositoryDurationSeconds, labelValues).Stop()
+		err := f(ctx)
+		if err != nil {
+			code := "unknown"
+			if e, ok := err.(errcode.Error); ok {
+				code = strings.ToLower(e.ErrorCode().String())
```

Comment:
what's the motivation for doing this?  Do we have cases where we report the same error code w/ different capitalization?

Reply:
I believe it makes sense at least in this case where [all the values are in UPPER CASE](https://github.com/dmage/image-registry/blob/dc3a1f03b58fc67d7e63e3607703a75b89692ec5/vendor/github.com/docker/distribution/registry/api/v2/errors.go#L15).

Reply:
Yes, I tried to avoid values in upper case to be consistent with other values, but removing ToLower to save CPU cycles is fine to me.